### ISSUE TITLE
Document how trial finds tests to run

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4526]
|<img alt="Screwtape's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| Screwtape reported|
|-|-|
|Trac ID|trac#4526|
|Type|enhancement|
|Created|2010-06-29 15:00:18Z|
|Branch|https://github.com/twisted/twisted/tree/trial-discovery-documentation-4526-2|

I've been on a bit of a trial documentation kick recently, and noticed that neither the [man page](http://twistedmatrix.com/documents/current/core/man/trial-man.xhtml) nor the [HTML docs](http://twistedmatrix.com/documents/current/core/howto/testing.html) discuss the algorithm trial uses when looking for test files.

For example, if I run:
```
trial my.python.package
```
then which of the following tests will be found?

 * `TestCase`s defined in `my/python/package/__init__.py`
 * `TestCase`s defined in `my/python/package/test_foo.py`
 * `TestCase`s defined in `my/python/package/Test_foo.py`
 * `TestCase`s defined in `my/python/package/testfoo.py`
 * `TestCase`s defined in `my/python/package/Testfoo.py`
 * `TestCase`s defined in `my/python/package/foo_test.py`

For freshly-written tests designed to work with Trial, I personally always use the `test_foo.py` variant, but sometimes when I want to use Trial's test-sniffing features with third-party libraries, or to test legacy code I've inherited, it would be nice to have a place where I could find out definitively what the rules are.

(yes, I suppose I could just run Trial and see, but sometimes it's easier to try nosetests or py.test or run things manually than decipher exactly what failure-mode caused a particular error message)

Attachments:

* [0001-Update-testing-with-how-Trial-finds-tests.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a7b/a7becf6c0b86ccff3320e0d9a9c70660a5761a34/89aab2eb6576ca5fe7f2ea14f4d0b3d7bb1f51fe.patch) (979 bytes) - added by edunham on 2014-04-14 18:24:21Z - 
* [document-how-trial-finds-tests.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a7b/a7becf6c0b86ccff3320e0d9a9c70660a5761a34/e006f09f1b2d746444df5120b0b4384103bec5db.patch) (906 bytes) - added by adamtheturtle on 2014-09-22 11:36:50Z - edunham's work with a NEWS file

<details><summary>Searchable metadata</summary>

```
trac-id__4526 4526
type__enhancement enhancement
reporter__Screwtape Screwtape
priority__normal normal
milestone__ 
branch__branches_trial_discovery_documentation_4526_2 branches/trial-discovery-documentation-4526-2
branch_author__adiroiban__rwall adiroiban, rwall
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__documentation_easy documentation easy
time__1277823618000000 1277823618000000
changetime__1422183607391800 1422183607391800
version__None None
owner__adiroiban adiroiban
cc__thijs cc__edunham
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @jml|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="arbiter's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|arbiter set owner to arbiter<br>arbiter set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="arbiter's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|arbiter removed owner<br>arbiter set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="edunham's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|edunham set owner to edunham<br>edunham set status to `assigned`|
|-|-|

The answer to this actually is [in the docs](http://twistedmatrix.com/documents/current/api/twisted.trial.runner.TestLoader.html#loadModule), just hidden pretty well. Whether a test is run depends not on what file it's in, but on how your module is configured and what its class inherits from. 

I guess that references to that documentation should be added to at least the [documentation](http://twistedmatrix.com/documents/current/core/howto/testing.html) and possibly the man page as well.

**Comment by twisted-trac** [comment]
|<img alt="edunham's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|edunham commented|
|-|-|

The man page links the howto page from "see also", so no changes to the man page are required. Attached above is a patch which adds a link to the module loader docs from the testing howto.

**Comment by twisted-trac** [comment]
|<img alt="edunham's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|edunham removed owner<br>edunham set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_10"></a>|@adiroiban set owner to edunham|
|-|-|

Thanks for the work on this.

The patch needs a news file. http://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles

The preferred format is diff without prefix http://twistedmatrix.com/trac/wiki/GitMirror#CreatingaGitclone

Other than that, it looks good.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="adamtheturtle's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|adamtheturtle removed owner|
|-|-|

Hopefully I am not stepping on anyone's toes picking up the ticket edunham has worked on. I have addressed adiroiban's comments by adding a patch with a NEWS file and I used the patch-making instructions from https://twistedmatrix.com/trac/wiki/BasicGuideToContributingCode#SubmitaPatch.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_13"></a>|@wallrj commented|
|-|-|

(In [43147]) Branching to trial-discovery-documentation-4526.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_14"></a>|@adiroiban commented|
|-|-|

All good! Please merge this! Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_15"></a>|@adiroiban commented|
|-|-|

(In [43735]) Branching to trial-discovery-documentation-4526-2.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_16"></a>|@adiroiban set owner to @adiroiban<br>@adiroiban set status to `new`|
|-|-|

I will sync with trunk and hope is ok... old branch was failing on buildbot

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_17"></a>|@adiroiban set status to `closed`|
|-|-|

(In [43737]) Merge trial-discovery-documentation-4526-2: Update trial documentation.

Author: adamtheturtle
Reviewers: adiroiban
Fixes: [#4526](https://github.com/twisted/twisted/issues/4526)

Update trial documentation with link to how trial discover tests.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun commented|
|-|-|

Don't forget to remove the review keyword when you review a ticket.

Also, the news fragment for this change seems to contain a mistake:

```
Twisted Trial's basics documentation now has a link to the documentation about how Trial finds tests.
Twisted Trial's basics documentation now has a link to the documentation about how Trial finds tests.
```

Separately, I'm not sure I'd consider this one line doc addition (which is just a reference to other existing documentation) non-misc material.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_19"></a>|@adiroiban commented|
|-|-|

hawkowl fixed this in [#7335](https://github.com/twisted/twisted/issues/7335) as the duplicate news line was already merged.
I am sorry for letting this pass the review and merged in trunk.

I will remove the review in the future.
Thanks for the help!
